Import badges

mahdibenromdhane opened the pull request.

ClaudeROSSI commented on src/scheduler/tasks/SynchronizeTagsImportTask.ts:
```diff
@@ -0,0 +1,75 @@
+                description: importedTag.description,
+                issuer: true,
+                deleted: false,
+                active: true,
```
yo,

active=true,
issuer=true

I cannot comment ... i do not know the impact of these properties ;)

ClaudeROSSI commented on src/assets/server/rest/v1/schemas/tag/tag-create-req.json:
```diff
@@ -0,0 +1,33 @@
+      "format": "date"
+    },
+    "createdBy": {
+      "type": "object"
```
?? why this is not a string?

mahdibenromdhane replied:
the createdBy is using the object format { id: userID }

LucasBrazi06 replied:
This schema is use againt data that will be saved in the DB
So there is no createdOn/By here

ClaudeROSSI reviewed (approved):
approved!
I only have questions to better understand some settings!
Thanks Mahdi

mahdibenromdhane commented:
> approved!
> I only have questions to better understand some settings!
> Thanks Mahdi

Ok :) Thanks

LucasBrazi06 commented on src/server/rest/v1/service/TagService.ts:
```diff
@@ -562,4 +643,24 @@ export default class TagService {
+        id: tag.id.toUpperCase(),
+        description: tag.description ? tag.description : `Tag ID '${tag.id}'`,
+        importedBy: req.user.id,
+        status: ImportStatus.UNKNOWN
```
Use a specific status like TO_BE_IMPORTED or NEWLY_IMPORTED that should be set when created in the collection

mahdibenromdhane replied:
removed the status since we have only two cases

LucasBrazi06 commented on src/server/rest/v1/service/UserService.ts:
```diff
@@ -1280,7 +1281,7 @@ export default class UserService {
         email: user.Email,
         importedBy: req.user.id,
-        status: UserImportStatus.UNKNOWN
+        status: ImportStatus.UNKNOWN
```
Same remark here.

LucasBrazi06 replied:
Rename First_Name -> FirstName

mahdibenromdhane replied:
Done

LucasBrazi06 commented on src/server/rest/v1/service/UserService.ts:
```diff
@@ -1280,7 +1281,7 @@ export default class UserService {
-        status: UserImportStatus.UNKNOWN
+        status: ImportStatus.UNKNOWN
       };
       await UserStorage.saveImportedUser(req.user.tenantID, newUploadedUser);
```
We should check the data with ajv (performant) before we save it in the DB

mahdibenromdhane replied:
Done

LucasBrazi06 commented on src/server/rest/v1/service/TagService.ts:
```diff
@@ -562,4 +643,24 @@ export default class TagService {
+        importedBy: req.user.id,
+        status: ImportStatus.UNKNOWN
+      };
+      await TagStorage.saveImportedTag(req.user.tenantID, newUploadedTag);
```
We should check the data with ajv (performant) before we save it in the DB

mahdibenromdhane replied:
Done

LucasBrazi06 commented on src/scheduler/tasks/SynchronizeTagsImportTask.ts:
```diff
@@ -0,0 +1,75 @@
+    const synchronizeTagsImport = LockingManager.createExclusiveLock(tenant.id, LockEntity.TAGS, 'synchronize-tags-import');
+    if (await LockingManager.acquire(synchronizeTagsImport)) {
+      try {
+        const importedTags = await TagStorage.getImportedTags(tenant.id, { statuses: [ImportStatus.UNKNOWN] }, Constants.DB_PARAMS_MAX_LIMIT);
```
If you have 1 million tags to import what do you think will happen to the VM's memory in prod?
Use pagination constants like EXPORT_RECORD_MAX_COUNT and process tags per block.
Moreover the count will be -1 in case of result > 500.

mahdibenromdhane replied:
Done

LucasBrazi06 commented on src/scheduler/tasks/SynchronizeTagsImportTask.ts:
```diff
@@ -0,0 +1,75 @@
+                issuer: true,
+                deleted: false,
+                active: true,
+                createdBy: { id: importedTag.importedBy },
```
I think you can assign right away the ID as a string:
`createdBy: importedTag.importedBy,`

LucasBrazi06 commented on src/scheduler/tasks/SynchronizeTagsImportTask.ts:
```diff
@@ -0,0 +1,75 @@
+                createdOn: new Date()
+              };
+              try {
+                TagValidator.getInstance().validateTagCreation(tag);
```
Should already be validated in Service endpoint.

mahdibenromdhane replied:
Done

LucasBrazi06 commented on src/scheduler/tasks/SynchronizeTagsImportTask.ts:
```diff
@@ -0,0 +1,75 @@
+                description: importedTag.description,
+                issuer: true,
+                deleted: false,
+                active: true,
```
Should be imported inactive

mahdibenromdhane replied:
Done

LucasBrazi06 commented on src/scheduler/tasks/SynchronizeUsersImportTask.ts:
```diff
@@ -19,12 +19,12 @@ export default class SynchronizeUsersImportTask extends SchedulerTask {
     const synchronizeUsersImport = LockingManager.createExclusiveLock(tenant.id, LockEntity.USER, 'synchronize-users-import');
```
Same remarks as the Tag import, no need to put back same comments

mahdibenromdhane replied:
Done

LucasBrazi06 commented on src/server/rest/v1/service/TagService.ts:
```diff
@@ -555,4 +636,24 @@ export default class TagService {
+    try {
+      const newUploadedTag: ImportedTag = {
+        id: tag.id.toUpperCase(),
+        description: tag.description ? tag.description : `Tag ID '${tag.id}'`,
```
Use i18n?
I think we use the term 'Badge' in the UI.

mahdibenromdhane replied:
Done and changed also for AddDescriptionToTagsTask

LucasBrazi06 commented on src/storage/mongodb/TagStorage.ts:
```diff
@@ -39,6 +39,118 @@ export default class TagStorage {
+    const tagMDB = {
+      _id: importedTagToSave.id,
+      description: importedTagToSave.description,
+      importedOn: new Date(),
```
We should reuse the import date linked to the import user and not creating a new one each time.

mahdibenromdhane replied:
Done

ClaudeROSSI replied:
Avoid the new Date() when updating an existing entity
ImportedOn: (importedTagToSave.ImportedOn)?importedTagToSave.ImportedOn:new Date()

LucasBrazi06 commented on src/storage/mongodb/TagStorage.ts:
```diff
@@ -39,6 +39,118 @@ export default class TagStorage {
+      { upsert: true, returnOriginal: false }
+    );
+    // Debug
+    await Logging.traceEnd(tenantID, MODULE_NAME, 'saveImportedTag', uniqueTimerID, tagMDB);
```
Should return the created tag ID for consistency reason like other API.

mahdibenromdhane replied:
Done

ClaudeROSSI replied:
ok

LucasBrazi06 commented on src/storage/mongodb/TagStorage.ts:
```diff
@@ -39,6 +39,118 @@ export default class TagStorage {
+        { 'description': { $regex: params.search, $options: 'i' } }
+      ];
+    }
+
```
Remove blank line

mahdibenromdhane replied:
Done

LucasBrazi06 commented on src/types/GlobalType.ts:
```diff
@@ -21,6 +21,12 @@ export interface Image {
+  UNKNOWN = 'U',
+  IMPORTED = 'I',
+  ERROR = 'E',
+}
```
If you remove UNKNOWN (initial upload) and IMPORTED (record is deleted), you don't need this enum anymore.
You will only have records which have no errors or ones with errors so only two cases in the end :-)
As you want.

mahdibenromdhane replied:
Done

LucasBrazi06 commented on src/types/Tag.ts:
```diff
@@ -14,3 +15,12 @@ export default interface Tag extends CreatedUpdatedProps {
+  description?: string;
+  importedBy: string;
+  importedOn?: Date;
+  error?: string;
```
Error should be an enum

mahdibenromdhane replied:
I removed this property and replaced it with errorCode / errorDescription

LucasBrazi06 commented on src/types/User.ts:
```diff
@@ -61,10 +62,10 @@ export interface ImportedUser {
-  status?: UserImportStatus
+  importedBy: string;
+  importedOn?: Date;
+  error?: string;
```
Error should be an enum

mahdibenromdhane commented on src/scheduler/tasks/SynchronizeTagsImportTask.ts:
```diff
@@ -0,0 +1,75 @@
+            const foundTag = await TagStorage.getTag(tenant.id, importedTag.id);
+            if (foundTag) {
+              importedTag.status = ImportStatus.ERROR;
+              importedTag.error = 'Tag id already exists';
```
Done

mahdibenromdhane commented on src/scheduler/tasks/SynchronizeTagsImportTask.ts:
```diff
@@ -0,0 +1,75 @@
+                TagValidator.getInstance().validateTagCreation(tag);
+                await TagStorage.saveTag(tenant.id, tag);
+                importedTag.status = ImportStatus.IMPORTED;
+                await TagStorage.saveImportedTag(tenant.id, importedTag);
```
Done

mahdibenromdhane commented on src/scheduler/tasks/SynchronizeTagsImportTask.ts:
```diff
@@ -0,0 +1,75 @@
+                });
+              } catch (error) {
+                importedTag.status = ImportStatus.ERROR;
+                importedTag.error = error.message;
```
Done

ClaudeROSSI commented on src/server/rest/v1/validator/TagValidation.ts:
```diff
@@ -0,0 +1,24 @@
+
+export default class TagValidator extends SchemaValidator {
+  private static instance: TagValidator|null = null;
+  private tagCreation: any;
```
put the type Schema here

import Schema from './Schema';

mahdibenromdhane replied:
Done

ClaudeROSSI commented on src/server/rest/v1/validator/TagValidation.ts:
```diff
@@ -0,0 +1,24 @@
+    return TagValidator.instance;
+  }
+
+  validateTagCreation(content): void {
```
content should be typed ? ImportedTag

mahdibenromdhane replied:
Done

ClaudeROSSI commented on src/server/rest/v1/service/TagService.ts:
```diff
@@ -458,6 +462,84 @@ export default class TagService {
+        void file.pipe(converter);
+      } else if (mimetype === 'application/json') {
+        const parser = JSONStream.parse('tags.*');
+        parser.on('data', async (tag) => {
+          await TagService.importTag(action, req, tag);
+        });
```
lint complains here:

![image](https://user-images.githubusercontent.com/26900551/111277763-1647e380-8639-11eb-9b7e-27727ff818da.png)

I'm not sure how to get rid of the warning and preserve the expected behavior!

Can you check without async/await, as shown below?

       const parser = JSONStream.parse('tags.*');
        parser.on('data', (tag) => {
          void TagService.importTag(action, req, tag);
        });

mahdibenromdhane replied:
Done

LucasBrazi06 commented on src/scheduler/tasks/SynchronizeTagsImportTask.ts:
```diff
@@ -0,0 +1,83 @@
+    const synchronizeTagsImport = LockingManager.createExclusiveLock(tenant.id, LockEntity.TAGS, 'synchronize-tags-import');
+    if (await LockingManager.acquire(synchronizeTagsImport)) {
+      try {
+        const dbParams: DbParams = { limit: Constants.EXPORT_PAGE_SIZE, skip: 0, onlyRecordCount: true };
+        let importedTags = await TagStorage.getImportedTags(tenant.id, { withNoError: true }, dbParams);
+        let count = importedTags.count;
+        delete dbParams.onlyRecordCount;
+        let skip = 0;
+        // Limit the number of records
+        if (count > Constants.EXPORT_RECORD_MAX_COUNT) {
+          count = Constants.EXPORT_RECORD_MAX_COUNT;
+        }
```
Yes, we should loop until there is no records remaining!
@mahdibenromdhane change the logic.

mahdibenromdhane replied:
We used this logic in the case that we don't update the retreived data. You'r right, it's not necessary in this case. 

Done.

ClaudeROSSI replied:
latest version is ok

ClaudeROSSI commented on src/scheduler/tasks/SynchronizeTagsImportTask.ts:
```diff
@@ -0,0 +1,77 @@
+    if (await LockingManager.acquire(synchronizeTagsImport)) {
+      try {
+        const dbParams: DbParams = { limit: Constants.EXPORT_PAGE_SIZE, skip: 0 };
+        let importedTags = await TagStorage.getImportedTags(tenant.id, { withNoError: true }, dbParams);
+        while (importedTags && !Utils.isEmptyArray(importedTags.result)) {
+          importedTags = await TagStorage.getImportedTags(tenant.id, { withNoError: true }, dbParams);
```
bug - twice the same call!

mahdibenromdhane replied:
Removed.

ClaudeROSSI commented on src/storage/mongodb/UserStorage.ts:
```diff
@@ -273,15 +273,15 @@ export default class UserStorage {
-      error: importedUserToSave.error,
+      errorCode: importedUserToSave.errorCode,
+      errorDescription: importedUserToSave.errorDescription,
       importedOn: new Date(),
```
same comment - preserve previous date on update!

mahdibenromdhane replied:
Done

ClaudeROSSI commented on src/scheduler/tasks/SynchronizeTagsImportTask.ts:
```diff
@@ -0,0 +1,75 @@
+
+export default class SynchronizeTagsImportTask extends SchedulerTask {
+  async processTenant(tenant: Tenant, config: TaskConfig): Promise<void> {
+    const synchronizeTagsImport = LockingManager.createExclusiveLock(tenant.id, LockEntity.TAGS, 'synchronize-tags-import');
```
@LucasBrazi06 - how do we proceed?
I think it makes sense to do it in a second PR.

ClaudeROSSI replied:
> We should also introduce the Async Task framework with this issue.
> We can do it in 2 steps.
> This job should delegate this to a task which can also be triggered by the Service after an import in the DB.

@LucasBrazi06 - how do we proceed?
I think it makes sense to do it in a second PR.

ClaudeROSSI reviewed (changes requested):
still some stuff to fix

ClaudeROSSI reviewed (approved):
ok.
@LucasBrazi06 : we couldn't start the Task Framework yet.

LucasBrazi06 commented:
> ok.
> @LucasBrazi06 : we couldn't start the Task Framework yet.

Okay, no problem.

LucasBrazi06 reviewed (changes requested):
- I refactored the file SynchronizeTagsImportTask.ts, check that it still works and apply the change I made in SynchronizeUsersImportTask.
- Also changed the TagService.handleImportTags() and importTag(), check it and report these changes in the UserService.
- I prefixed the schema names with 'imported' to avoid collision with real creation of a User and a Tag
- I renamed the error code to status: either the badge/user is imported and ready to be processed by the job or there is an error and we have in the end 2 statuses: READY and ERROR. I created an enum called ImportStatus used by User and Tag and renamed errorCode by status and adjusted the Service and Storage class. No HttpStatus in data type, this is HTTP code used at high level service interface!
-

mahdibenromdhane commented:
> * I refactored the file SynchronizeTagsImportTask.ts, check that it still works and apply the change I made in SynchronizeUsersImportTask.
> * Also changed the TagService.handleImportTags() and importTag(), check it and report these changes in the UserService.
> * I prefixed the schema names with 'imported' to avoid collision with real creation of a User and a Tag
> * I renamed the error code to status: either the badge/user is imported and ready to be processed by the job or there is an error and we have in the end 2 statuses: READY and ERROR. I created an enum called ImportStatus used by User and Tag and renamed errorCode by status and adjusted the Service and Storage class. No HttpStatus in data type, this is HTTP code used at high level service interface!

Okey

LucasBrazi06 commented:
@mahdibenromdhane I havent't had time to test my changes in importing Tags. You can start here :). Thanks!

mahdibenromdhane commented:
> @mahdibenromdhane I havent't had time to test my changes in importing Tags. You can start here :). Thanks!

ok, I will test and do the necessary changes :)

jerome-benoit commented:
> @mahdibenromdhane I havent't had time to test my changes in importing Tags. You can start here :). Thanks!
A more efficient code flow would have been to do a PR on top of that branch. Developers and reviewers would have a clear view of the changes made, been able to review them, not been forced to accept them, learnt by seeing clearly all the changes, been involved in the review process and made responsible over the whole work. 
Furthermore, it's impolite to make changes in an other developer work and branch without asking him first.

LucasBrazi06 commented:
> > @mahdibenromdhane I havent't had time to test my changes in importing Tags. You can start here :). Thanks!
> 
> A more efficient code flow would have been to do a PR on top of that branch. Developers and reviewers would have a clear view of the changes made, been able to review them, not been forced to accept them, learnt by seeing clearly all the changes, been involved in the review process and made responsible over the whole work.
> Furthermore, it's impolite to make changes in an other developer work and branch without asking him first.

You should mind your business and let me drive my team as I want, that's not your role.
And it's even more impolite to make this kind of statement in the GitHub repo.
I'll delete it and next time you'll meet Hanno.

ClaudeROSSI commented on src/server/rest/v1/service/TransactionService.ts:
```diff
@@ -1013,8 +1013,7 @@ export default class TransactionService {
 
-  private static async getTransactions(req: Request, action: ServerAction,
-    params: { completedTransactions?: boolean, withTag?: boolean } = {}, projectFields): Promise<DataResult<Transaction>> {
+  private static async getTransactions(req: Request, action: ServerAction, params: { withTag?: boolean } = {}, projectFields): Promise<DataResult<Transaction>> {
```
?? suspicious change. right?

mahdibenromdhane replied:
Error changes due to merge conflict.
Fixed

ClaudeROSSI commented on src/server/rest/v1/service/UserService.ts:
```diff
@@ -19,7 +20,7 @@ import Cypher from '../../../../utils/Cypher';
 import EmspOCPIClient from '../../../../client/ocpi/EmspOCPIClient';
 import I18nManager from '../../../../utils/I18nManager';
-import JSONStream from 'JSONStream';
+import JSONStream from 'jsonstream';
```
jsonstream lowercase here but uppercase in TagService. Which one is the correct one?
please double check.
Thanks

mahdibenromdhane replied:
The uppercase is correct, typo while trying other modules. 
Fixed

mahdibenromdhane commented on src/types/Tag.ts:
```diff
@@ -24,3 +24,6 @@ export interface ImportedTag {
   errorDescription?: string;
 }
+
+export const TagRequiredImportProperties = ['id'];
```
We are checking both client and server side, 
I'v missed the server implementation
Fixed

ClaudeROSSI reviewed (approved):
approved
